src/clients/KubectlClient.ts

```diff
@@ -101,10 +100,73 @@ export class KubectlClient implements IClient {
+                    .filter(s => !knownSideCars.find(knownSideCar => knownSideCar == s));
+            } else {
+                this._logger.error(TelemetryEvent.KubectlClient_GetContainerListError, new Error(`container list is null`));
+                return null;
+            }
+        } catch (error) {
+            this._logger.error(TelemetryEvent.KubectlClient_GetContainerListError, error);
+            return null;
+        }
+
+    }
+
+    public async getPodName(serviceName: string): Promise<string> {
+        try {
+            const kubeconfigPath: string = await this._accountContextManager.getKubeconfigPathAsync();
+            // find the pod ipaddressList for the selected service
+            const args: string[] = [`get`, `ep`, serviceName, `-o`, `jsonpath='{.subsets[*].addresses[*].ip}'`];
+            let output = await this.runKubectlCommandAsync(args, kubeconfigPath);
+            // if there replicas and multiple pods, split it by apostrope, space example value: ''10.2.45.6 10.5.6.809''
+            // handles single value as well example value: ''10.56.78.90''
+            let ipaddressList: string[];
+            if (output && output.length > 0) {
+                ipaddressList = output.replace(/'/g, '').replace(/[\s,]+/g, ',').split(',');
+            }
+
+            if (ipaddressList?.length > 0) {
+                // find the podname by looping the ip address
+                for (const ip in ipaddressList) {
+                    const args2: string[] = [`get`, `pods`];
+                    const fieldSelector = '--field-selector=status.podIP='.concat(ipaddressList[ip]);
+                    args2.push(fieldSelector);
+                    args2.push("-o=name");
+                    const finalOutput = await this.runKubectlCommandAsync(args2, kubeconfigPath);
+                    // split the output by forward slash and new line char at the end example value: 'pod/stats-api-ff7d66c5b-4nc9x\n' 
+                    // output is stats-api-ff7d66c5b-4nc9x
+                    if (finalOutput?.length > 0) {
+                        return finalOutput.indexOf("/") != -1 ? finalOutput.split("/")[1].split("\n")[0] : finalOutput.split("\n")[0];
+                    }
+                }
+            }
+            return null;
+        } catch (error) {
+            this._logger.error(TelemetryEvent.KubectlClient_GetPodNameError, error);
+            // not throwing error to continue the flow
+            return null;
+        }
+    }
+
     public async getServicesAsync(namespace: string = null): Promise<IKubernetesService[]> {
         const kubeconfigPath: string = await this._accountContextManager.getKubeconfigPathAsync();
-        const args: string[] = [ `get`, `services`, `-o`, `json` ];
-        args.push(...(namespace != null ? [ `-n`, namespace ] : [ `--all-namespaces` ]));
+        const args: string[] = [`get`, `services`, `-o`, `json`];
```

Comment:
I've managed to thoroughly confuse myself in the past trying to remember how to interpret namespace being `null` or `undefined`. I see it means "all namespaces" here, but it could just as easily mean "default namespace" (as `kubectl` would understand it). This may be a job for a different PR, but I'd quite like to see namespace represented by a dedicated TypeScript type where the meaning is made explicit, and you can't accidently interpret it wrongly.

Reply:
I am refactoring all this with javascript client for k8s. this will be replaced